File: common/src/main/java/generations/gg/generations/core/generationscore/common/compat/jei/RksRecipeExtension.java
package generations.gg.generations.core.generationscore.common.compat.jei;

import generations.gg.generations.core.generationscore.common.world.recipe.GenerationsIngredient;
import generations.gg.generations.core.generationscore.common.world.recipe.RksRecipe;
import generations.gg.generations.core.generationscore.common.world.recipe.ShapedRksRecipe;
import mezz.jei.api.gui.builder.IRecipeLayoutBuilder;
import mezz.jei.api.gui.ingredient.ICraftingGridHelper;
import mezz.jei.api.recipe.IFocusGroup;
import mezz.jei.library.plugins.vanilla.crafting.JeiShapedRecipe;
import mezz.jei.library.util.RecipeUtil;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.crafting.CraftingRecipe;
import net.minecraft.world.item.crafting.RecipeHolder;
import net.minecraft.world.item.crafting.ShapedRecipe;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class RksRecipeExtension<T extends RksRecipe> implements IRksCategoryExtension<T> {
    @Override
    public void setRecipe(RecipeHolder<T> recipeHolder, IRecipeLayoutBuilder builder, ICraftingGridHelper craftingGridHelper, IFocusGroup focuses) {
        RksRecipe recipe = recipeHolder.value();
        ItemStack resultItem = RecipeUtil.getResultItem(recipe);

        int width = getWidth(recipeHolder);
        int height = getHeight(recipeHolder);
        craftingGridHelper.createAndSetOutputs(builder, List.of(resultItem));
        craftingGridHelper.createAndSetIngredients(builder, recipe.getIngredients(), width, height);
    }

    @SuppressWarnings("removal")
    @Override
    public Optional<ResourceLocation> getRegistryName(RecipeHolder<T> recipeHolder) {
        return Optional.of(recipeHolder.id());
    }

    @Override
    public int getWidth(RecipeHolder<T> recipeHolder) {
        T recipe = recipeHolder.value();
        if (recipe instanceof ShapedRksRecipe shapedRecipe) {
            return shapedRecipe.getWidth();
        }

        return 0;
    }

    @Override
    public int getHeight(RecipeHolder<T> recipeHolder) {
        T recipe = recipeHolder.value();
        if (recipe instanceof ShapedRksRecipe shapedRecipe) {
            return shapedRecipe.getHeight();
        }
        return 0;
    }

    @Override
    public boolean isHandled(RecipeHolder<T> recipe) {
        return !recipe.value().isSpecial();
    }
}

File: src/common/main/kotlin/generations/gg/generations/core/generationscore/common/compat/jei/RksRecipeExtension.kt
//package generations.gg.generations.core.generationscore.common.compat.jei
//
//import generations.gg.generations.core.generationscore.common.client.render.rarecandy.instanceOrNull
//import generations.gg.generations.core.generationscore.common.world.recipe.RksRecipe
//import generations.gg.generations.core.generationscore.common.world.recipe.ShapedRksRecipe
//import mezz.jei.api.gui.builder.IRecipeLayoutBuilder
//import mezz.jei.api.gui.ingredient.ICraftingGridHelper
//import mezz.jei.api.recipe.IFocusGroup
//import mezz.jei.library.util.RecipeUtil
//import net.minecraft.resources.ResourceLocation
//import net.minecraft.world.item.crafting.RecipeHolder
//import java.util.*
//import java.util.List
//
//class RksRecipeExtension<T : RksRecipe> : IRksCategoryExtension<T> {
//    override fun setRecipe(
//        recipeHolder: RecipeHolder<T>,
//        builder: IRecipeLayoutBuilder,
//        craftingGridHelper: ICraftingGridHelper,
//        focuses: IFocusGroup
//    ) {
//        val recipe: RksRecipe = recipeHolder.value()
//        val resultItem = RecipeUtil.getResultItem(recipe)
//
//        val width = getWidth(recipeHolder)
//        val height = getHeight(recipeHolder)
//        craftingGridHelper.createAndSetOutputs(builder, List.of(resultItem))
////        craftingGridHelper.createAndSetIngredients(builder, recipe.ingredients, width, height)
//    }
//
//    override fun getWidth(recipeHolder: RecipeHolder<T>): Int {
//        return recipeHolder.value().instanceOrNull<ShapedRksRecipe>()?.pattern?.width ?: 0
//    }
//
//    override fun getHeight(recipeHolder: RecipeHolder<T>): Int {
//       return recipeHolder.value().instanceOrNull<ShapedRksRecipe>()?.pattern?.height ?: 0
//    }
//
//    override fun isHandled(recipe: RecipeHolder<T>): Boolean {
//        return !recipe.value().isSpecial
//    }
//}
